### PR TITLE
SI-4788/SI-5948 Respect RetentionPolicy of Java annotations

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

**Comment by soc** [comment]
Btw, I eliminated GenJVMASM completely and moved stuff used by both GenBCode and GenASM to BCodeASMCommon.

File: src/compiler/scala/tools/nsc/symtab/SymbolLoaders.scala [hunk 1]
```diff
@@ -206,7 +206,7 @@ abstract class SymbolLoaders {
 
     override def complete(root: Symbol) {
       try {
-        val start = currentTime
+        val start = System.nanoTime()
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/symtab/SymbolLoaders.scala]: What's the difference?

Comment (soc) [inline src/compiler/scala/tools/nsc/symtab/SymbolLoaders.scala]: Argh, looks like an oversight from a different commit.
I'll put this and the other change into a separate change.

The difference is that `System.nanoTime` measures elapsed time (which is guaranteed to only increase monotonically on most platforms in a thread (thanks Microsoft for messing this up)), while `System.currentTimeMillis` tracks time since epoch (which can easily jump forward and backward).

**Comment by soc** [comment]
Ok, fixed. Btw, do I need both a test for combined and separate compilation, or is separate compilation one enough?

**Comment by retronym** [comment]
Please include both. I doubt there will be a difference here, usually that only happens if you touch Unpickler and forget to change JavaParsers, or vice versa. Perhaps it can also show up a missing call to Symbol#initialize.

**Comment by soc** [comment]
Ok, great, both tests are already included.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala [hunk 2]
```diff
@@ -124,4 +131,29 @@ final class BCodeAsmCommon[G <: Global](val global: G) {
...
+
+  def isRuntimeVisible(annot: AnnotationInfo): Boolean =
+    annot.atp.typeSymbol.getAnnotation(AnnotationRetentionAttr)
+      .exists(_.assocs.contains((nme.value -> LiteralAnnotArg(Constant(AnnotationRetentionPolicyRuntimeValue)))))
+
+  def retentionPolicyOf(annot: AnnotationInfo): Symbol =
+    annot.atp.typeSymbol.getAnnotation(AnnotationRetentionAttr).map(_.assocs).map(assoc =>
+      assoc.collectFirst {
+        case (`nme`.value, LiteralAnnotArg(Constant(value: Symbol))) => value
+      }.getOrElse(AnnotationRetentionPolicyClassValue)).getOrElse(AnnotationRetentionPolicyClassValue)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala]: Looks like you could do `.flatten.getOrElse(AnnotationRetentionPolicyClassValue)` instead of putting the default twice

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala [hunk 3]
```diff
@@ -124,4 +131,29 @@ final class BCodeAsmCommon[G <: Global](val global: G) {
...
+    * .initialize: if 'annot' is read from pickle, atp might be un-initialized
+    */
+  def shouldEmitAnnotation(annot: AnnotationInfo) = {
+    annot.symbol.initialize.isJavaDefined &&
+      annot.matches(ClassfileAnnotationClass) &&
+      retentionPolicyOf(annot) != AnnotationRetentionPolicySourceValue &&
+      annot.args.isEmpty
+  }
+
+  def isRuntimeVisible(annot: AnnotationInfo): Boolean =
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala:150]: Let's keep this and `retentionPolicyOf` private.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala (hunk 2 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala]: I wonder if we should switch over to deconstructing `AnnotationInfo.tree` now.

``` scala
scala> class C { @scala.reflect.ScalaLongSignature(bytes = Array[Byte]()) def foo = "" }
defined class C
warning: previously defined object C is not a companion to class C.
Companions must be defined together; you may wish to use :paste mode for this.

scala> val foo = typeOf[C].member(TermName("foo"))
foo: $r.intp.global.Symbol = method foo

scala> val annot = foo.annotations.head
annot: $r.intp.global.AnnotationInfo = scala.reflect.ScalaLongSignature(bytes = [])

scala> val atp = annot.atp
atp: $r.intp.global.Type = scala.reflect.ScalaLongSignature

scala> val treeInfo.Applied(_, _, ((AssignOrNamedArg(Ident(`nme`.value), Literal(Constant(cls: Symbol))) :: Nil) :: Nil)) = atp.typeSymbol.initialize.getAnnotation(symbolOf[java.lang.annotation.Retention]).get.tree
cls: $r.intp.global.Symbol = value RUNTIME
```

I wonder if this we could also try to extract some generally useful combinator from this and implement in terms of that.

File: test/files/run/t4788/Test.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      .find(_.contains("@L"))
+      .map(_.trim)
+  }
+
+  def show {
+    // It seems like @java.lang.Deprecated shows up in both the
+    // Deprecated attribute and RuntimeVisibleAnnotation attribute,
+    // while @scala.deprecated only shows up in the Deprecated attribute.
+    // The check file just documents status quo, not sure if Scala
+    // should brought in line with Java or not...
```
Comment (lrytz) [inline test/files/run/t4788/Test.scala:26]: I looked a bit how we handle `@deprecated`, `@Deprecated`, java, scala, etc, it seems things are not consistent currently. This will require a cleanup, but it does not need to happen in this PR. I filed https://issues.scala-lang.org/browse/SI-8883

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala (hunk 2 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala]: I'm fine with both.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala [hunk 5]
```diff
@@ -23,6 +23,8 @@ abstract class BCodeHelpers extends BCodeIdiomatic with BytecodeWriters {
   import bTypes._
   import coreBTypes._
 
+  val bCodeAsmCommon: BCodeAsmCommon[global.type] = new BCodeAsmCommon(global)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala]: There exists already an instance of `BCodeAsmCommon`, it's `bTypes.bCodeAsmCommon`.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala [hunk 6]
```diff
@@ -469,6 +471,7 @@ abstract class BCodeHelpers extends BCodeIdiomatic with BytecodeWriters {
   trait BCAnnotGen extends BCInnerClassGen {
 
     import genASM.{ubytesToCharArray, arrEncode}
+    import bCodeAsmCommon._
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala]: just import `bCodeAsmCommon.isRuntimeVisible`.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala [hunk 7]
```diff
@@ -124,4 +131,29 @@ final class BCodeAsmCommon[G <: Global](val global: G) {
...
+  lazy val AnnotationRetentionPolicyRuntimeValue = AnnotationRetentionPolicyModule.tpe.member(TermName("RUNTIME"))
+
+  /** Whether an annotation should be emitted as a Java annotation
+    * .initialize: if 'annot' is read from pickle, atp might be un-initialized
+    */
+  def shouldEmitAnnotation(annot: AnnotationInfo) = {
+    annot.symbol.initialize.isJavaDefined &&
+      annot.matches(ClassfileAnnotationClass) &&
+      retentionPolicyOf(annot) != AnnotationRetentionPolicySourceValue &&
+      annot.args.isEmpty
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala:147]: There used to be `!annot.matches(definitions.DeprecatedAttr)` – are you sure removing it doesn't change behavior?

File: test/files/run/t4788.check [hunk 8]
```diff
@@ -0,0 +1,6 @@
+warning: there was one deprecation warning; re-run with -deprecation for details
```
Comment (lrytz) [inline test/files/run/t4788.check]: why do we get a deprecation warning here?

**Comment by ericacm** [comment]
Does this mean @Retention(RUNTIME) annotations are going to be supported?

**Comment by gkossakowski** [comment]
I'd like to include this in 2.11.3.
@soc: Let me know if you can address review feedback by tomorrow's evening.

File: test/files/run/t4788/Test.scala (hunk 4 above)
Comment (soc) [inline test/files/run/t4788/Test.scala:26]: I commented on the issue tracker.

File: test/files/run/t4788.check (hunk 8 above)
Comment (soc) [inline test/files/run/t4788.check]: Because of

```
@deprecated
class DScala
```

In the tests. I felt that `@deprecated` and `@Deprecated` where important/special-cased enough to document their current behavior explicitly.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala (hunk 7 above)
Comment (soc) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala:147]: See the commit message:

```
Note that I removed the check to ignore @deprecated:
- @deprecated extends StaticAnnotation, so they aren't
  supposed to show up in the RuntimeInvisibleAnnotation
  attribute anyway, and the earlier check for "extends
  ClassfileAnnotationClass" makes this check superflous
  anyway.
- Otherwise, if @deprecated was extending
  ClassfileAnnotationClass it would seem inconsistent
  that we don't emit @deprecated, but would do so for
  @deprecatedOverriding, @deprecatedInheritance, etc.
  Anyway, due to ClassfileAnnotation not working in
  Scala, and the additional check which only allows
  Java-defined annotations, this is pretty pointless
  from every perspective.
```

**Comment by soc** [comment]
@gkossakowski I'm on it, it will just take a bit because @scala-ide grinds to a halt after ... saving. So I basically have only one save before I have to kill -9 Eclipse and restart it.

**Comment by gkossakowski** [comment]
Ok, no need to rush, though. The next Scala release is planned for the end of November.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala (hunk 3 above)
Comment (soc) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala:150]: I made `retentionPolicyOf` private, but `isRuntimeVisible` cannot be made private because it is used in `GenASM` and `BCodeHelpers`.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala (hunk 6 above)
Comment (soc) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala]: Thanks!

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala (hunk 2 above)
Comment (soc) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala]: I've picked the `.flatten.getOrElse(AnnotationRetentionPolicyClassValue)` for now.

**Comment by soc** [comment]
Btw, when did this warning show up:

```
[quick.compiler] /home/soc/Entwicklung/scala/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:616: warning: matches with two cases or fewer are emitted using if-then-else instead of switch
[quick.compiler]               (argsSize : @switch) match {
[quick.compiler]                            ^
```

**Comment by gkossakowski** [comment]
Today, see #4016. It looks like we'll need to clean up our own code base after #4016 got merged.

**Comment by soc** [comment]
@ericacm It's a bit more complicated. What this commit does is that it fixes the visibility of Java annotations marked with `RetentionPolicy.CLASS`. Before this commit, annotations with `RetentionPolicy.CLASS` would show up as if they were decalred with `RetentionPolicy.RUNTIME`.

Annotations with `RetentionPolicy.RUNTIME` have been supported correctly since at least Scala 2.7.x as far as I remember.

Note that the above only applies to annotations defined in Java. This commit doesn't change how Scala annotations are handled.

I'd like to fix `ClassfileAnnotation` and add `RuntimeAnnotation` so that annotations with these visibility levels can be defined in Scala, too, but I have no idea what would be required here.

Basically, the tree of the annotation declaration would need to be transformed into the thing the JVM bytecode expects, and I honestly have no idea what that is and especially how close that is to the thing scalac does already.

Maybe this has become a bit easier with the ASM backend? I'll have to look whether they have some special methods to emit annotations declarations or whether it has be done by emitting classes/interfaces.

**Comment by soc** [comment]
@gkossakowski Done. If @retronym and @lrytz are happy with the changes, I think this is good to go.

**Comment by gkossakowski** [comment]
@retronym has relocated to Australia so we'll to wait a few more hours before he wakes up.

File: test/files/run/t4788/Test.scala (hunk 4 above)
Comment (lrytz) [inline test/files/run/t4788/Test.scala:26]: thanks!

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala (hunk 7 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala:147]: Thanks, I missed the commit message

File: test/files/run/t4788.check (hunk 8 above)
Comment (lrytz) [inline test/files/run/t4788.check]: I don't understand - as far as I can see, `DScala` is declared, but never referenced (only reflectively in a  String). That should not trigger a deprecation warning, no? Also, there seems to be only a deprecation warning in `t4788.check`, but not in `t4788-separate-compilation.check`.

**Comment by lrytz** [comment]
My only open question is https://github.com/scala/scala/pull/4026#discussion_r18448760, otherwise this looks all good!

File: test/files/run/t4788.check (hunk 8 above)
Comment (soc) [inline test/files/run/t4788.check]: Argh, damn. You are right. The deprecation message was caused by me not using the API correctly:

```
D.scala:4: warning: @deprecated now takes two arguments; see the scaladoc.
@deprecated
 ^
one warning found
```

Comment (soc) [inline test/files/run/t4788.check]: Not sure about the difference between combined and separate compilation. Maybe warnings from earlier compilation runs are not reported?

Comment (soc) [inline test/files/run/t4788.check]: Warning is gone.

**Comment by lrytz** [comment]
LGTM! Not sure why :pouting_cat: is assigning red labels, hopefully she'll start a build soon enough.

**Comment by gkossakowski** [comment]
@lrytz, you win today's emoji contest. I'll :point_right: my way out.

**Comment by gkossakowski** [comment]
@soc: it seems like the scope of this PR got narrowed down. Could you update PR's title and description to match what has been merged?

**Comment by soc** [comment]
@gkossakowski I'd do that, but I have no idea where the diff between the change and the title/description is ...

**Comment by retronym** [comment]
Great result!

@soc While you're conquering the world of annotations, are you also planning to take a swing at [SI-4242](https://issues.scala-lang.org/browse/SI-4242), support for `java.lang.annotation.Target`?

**Comment by lrytz** [comment]
@retronym for SI-4242, one question to discuss is if we want a similar mechanism for Scala annotations (could be integrated with our existing target annotations). Then we would probably map Java Target annotations to that.

**Comment by retronym** [comment]
Yep, I think we'd want feature parity.

Another related request was for the ability to put a primary constructor annotation outside the class:

``` scala
(@(Foo @target.primaryConstructor)) class C

// as an alternative to the clunky looking
class C @Foo()
```

**Comment by ericacm** [comment]
Thanks for the clarification.

I would like to be able to declare Scala classes with
RetentionPolicy.RUNTIME for JSR-303 validation.  The workaround now is to
use Java classes, just like you have to do for Java-friendly enums.

On Tuesday, October 7, 2014, soc notifications@github.com wrote:

> @ericacm https://github.com/ericacm It's a bit more complicated. What
> this commit does is that it fixes the visibility of Java annotations marked
> with RetentionPolicy.CLASS. Before this commit, annotations with
> RetentionPolicy.CLASS would show up as if they were decalred with
> RetentionPolicy.RUNTIME.
> 
> Annotations with RetentionPolicy.RUNTIME have been supported correctly
> since at least Scala 2.7.x as far as I remember.
> 
> Note that the above only applies to annotations defined in Java. This
> commit doesn't change how Scala annotations are handled.
> 
> I'd like to fix ClassfileAnnotation and add RuntimeAnnotation so that
> annotations with these visibility levels can be defined in Scala, too, but
> I have no idea what would be required here.
> 
> Basically, the tree of the annotation declaration would need to be
> transformed into the thing the JVM bytecode expects, and I honestly have no
> idea what that is and especially how close that is to the thing scalac does
> already.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/scala/scala/pull/4026#issuecomment-58206931.

## 

Sent from Gmail Mobile

**Comment by ericacm** [comment]
That would be very cool, although that syntax actually also looks pretty
darn klunky :)

But hopefully you can clean it up like

type FooConst = Foo @target.primaryConstructor

@FooConst class C

This approach currently works for other targeting annotations.

-- Eric

**Comment by retronym** [comment]
Yep, that would work.

**Comment by soc** [comment]
@retronym I was planning to look at SI-4242 for a long time already, but usually other issues were higher on my priority list, so this one lost. Feel free to steal it if you want, otherwise I'll take a look at it again.

I think coming up with a good design which interops with Java's target annotations while not creating more hard dependencies between Scala and Java is the harder part.

Maybe we can type-alias Java's existing target annotations and just add a few of our own which are missing?

**Comment by soc** [comment]
@retronym I agree with the constructor issue. Play recently introduced constructor dependency injection, and it looks just terrible currently: https://www.playframework.com/documentation/2.4.x/ScalaDependencyInjection

```
class MyComponent @Inject() (ws: WSClient) {
  // ...
}
```

**Comment by soc** [comment]
@ericacm Support for JVM annotations and JVM enums are both on my list of things I'd like to do.
Actually, enums are pretty far along the road already, except an issue I can't really figure out.

Edit: Here is the link: https://groups.google.com/forum/#!topic/scala-internals/LF9A6-UR7yM
